test/jdk/java/lang/StableValue/JEP.md

```diff
@@ -0,0 +1,561 @@
+| Mutable            | [0, ∞)   | Anywhere                          |
+| `final`            | 1        | Constructor or static initializer |
+| at-most-once (N/A) | [0, 1]   | Anywhere                          |
+
+"At-most-once fields" would be essential to expensive cache computations associated with method calls, so that they can
+be reused across multiple calls. For instance, creating a logger or reading application configurations from an external
+database. Furthermore, if the VM is made aware, a field is an "at-most-once field" and it is set, it may
+[constant-fold](https://en.wikipedia.org/wiki/Constant_folding) the field value, thereby providing crucial performance
+and energy efficiency gains. 
+
+It is also important to stress a method called to compute a value might have intended or
+unintended side effects and therefore, it would be vital to also guarantee the method is invoked at most once, even in
+a multithreaded environment.
+
+Another property of at-most-once fields would be, they are written to at most once but
+are likely read at many occasions. Hence, updating the field is not so time-critical whereas every effort to make 
+reading the field performant should be made.  
+
+Here is how a naïve cache could look like using a mutable field and where a `Logger` instance is cached:
+
+```
+// A naïve cache. Do not use this solution!
+public class Cache {
+
+    private Logger logger;
+
+    public Logger logger() {
+        Logger v = logger;
+        if (v == null) {
+            logger = v = Logger.getLogger("com.company.Foo");
+        }
+        return v;
+    }
+}
+```
+
+This solution does not work in a multithreaded environment as updates made by one thread to the `logger` may not be
+seen by other threads, thereby allowing the `logger` variable to be updated several times and consequently the
+`Logger::getLogger` method can be called several times.
+
+Here is how thread safety can be added together with a guarantee, `logger` is only updated at most once
+(and accordingly `Logger::getLogger` is called at most once):
+
+```
+// A field protected by synchonization. Do not use this solution!
+public class Cache {
+
+    private Logger logger;
+
+    public synchronized Logger logger() {
+        Logger v = logger;
+        if (v == null) {
+            logger = v = Logger.getLogger("com.company.Foo");
+        }
+        return v;
+    }
+}
+```
+
+While this works, acquiring the `synhronized` monitor is slow and prevents multiple threads from accessing the cached
+`Logger` instance simultaneously once computed.
+
+The solution above can be modified to use the
+[*double-checked locking idiom*](https://en.wikipedia.org/wiki/Double-checked_locking) which would improve the
+situation a bit:
+
+```
+// A field protected by double-checked locking. Do not use this solution!
+public class Cache {
+
+    private volatile Logger logger;
+
+    public Logger logger() {
+        Logger v = logger;
+        if (v == null) {
```

Comment:
You should comment that reading and writing logger to a local variable is an optimization trick; it's still correct when `logger` is just read twice or write-and-read because of the volatile-imposed ordering, but a local var saves the penalty of a volatile read.